Prepare v1, looking for feedback

cometkim opened the pull request:
Looking for feedback for alpha release:

```bash
npm install --save gatsby-plugin-typegen@alpha

# or

yarn add gatsby-plugin-typegen@alpha
```

supported options: https://github.com/cometkim/gatsby-plugin-typegen/blob/v1/src/types.ts


### Tasks

- [x] Resolves #13 - Nees some feedback.
- [x] Resolves #14 
- [x] Resolves #19 - Need test. 
- [x] Resolves #20 - Replace all console.log to the reporter, and use verbose so the plugin is quiet by default and users will be able to enable logs explicitly with the `--verbose` flag.
- [x] Resolves #21
- [x] Resolves #26 
- [x] Closes #18 
- [x] Update README

### Additional tasks

- Roadmap for next release -> #27 
- Add tests and examples - Done with a manual test, will add some examples so can it be checked automatically via CI -> #28 #29

cometkim committed:
WIP: v1

cometkim committed:
Extract more utilities for gatsby from original impl

cometkim committed:
Alpha 1

cometkim committed:
Validate resolvers plugin is installed correctly

cometkim committed:
Completly rewritten

cometkim committed:
Add missing(implicitly used) dependecies

cometkim committed:
Cleanup unused deps

cometkim committed:
Move local type declarations

cometkim committed:
Alpha 3

cometkim committed:
Remove files whitelist, add npmignore

cometkim committed:
v0.3.0-alpha.4

cometkim committed:
Remove graphql from deps, use Gatsby's one instead

cometkim committed:
v0.3.0-alpha.5

cometkim committed:
Fix documents loader

cometkim committed:
Add missing dependencies

cometkim committed:
v0.3.0-alpha.6

cometkim committed:
Cleanup codes

cometkim commented on package.json:
```diff
@@ -11,47 +11,49 @@
+    "has-yarn": "^2.1.0",
     "normalize-path": "^3.0.0"
   },
+  "optionalDependencies": {
```
:thinking: I'm still not sure this actually make sense

d4rekanguok replied:
I think you'd still need to include these in `dependencies`. AFAIK `optionalDependencies` implies that these dependencies is not required _at all_, and the library will handle that case internally. For us, if a user set `includeResolvers: true` and the `@graphql-codegen/typescript-resolvers` happened to be missing from node_module, the lib will fail.

d4rekanguok replied:
Ah I've just seen the handling code below — I guess that make sense, but IMHO it also make the code unnecessary complex :P

cometkim commented on src/workers/insert-types.ts:
```diff
@@ -0,0 +1,81 @@
+      .replace(STATIC_QUERY_HOOK_REGEXP, STATIC_QUERY_HOOK_REPLACER)
+      .replace(STATIC_QUERY_COMPONENT_REGEXP, STATIC_QUERY_COMPONENT_REPLACER)
+
+    if (content.length !== fixed.length) {
```
```suggestion
    if (content !== fixed) {
```

This is ok.

cometkim commented on src/workers/codegen.ts:
```diff
@@ -0,0 +1,97 @@
+      codegenOptions.pluginMap['typescriptOperations'] = require('@graphql-codegen/typescript-operations');
+      codegenOptions.plugins.push({ typescriptOperations: DEFAULT_TYPESCRIPT_CONFIG });
+      if (includeResolvers) {
+        codegenOptions.plugins.push({ typescriptResolvers: {} });
```
Export `GatsbyResolverContext`, so help people build custom schema with types

d4rekanguok commented on src/gatsby-node.ts:
```diff
@@ -0,0 +1,129 @@
+import path from 'path';
+import chokidar from 'chokidar';
+import { printSchema } from 'graphql';
```
I think it'd be better to import graphql from gatsby directly i.e
```
import { printSchema } from 'gatsby/graphql'
```

d4rekanguok commented on src/plugin-utils.ts:
```diff
@@ -0,0 +1,129 @@
+
+  const {
+    language = 'typescript',
+    includeResolvers = false,
```
Would it be better to let user specify additional @graphql-codegen plugin themselves, instead of doing a toggle here? i.e If someone needs the resolver plugin, they could require it in gatsby-config then add it to a 'codegenPlugins' array

cometkim replied:
I thought about project goals. It's not about writing a codegen configuration preset, but about integrating development with types to be more natural in Gatsby.

I think hiding codegen interfaces from the user would be better.

So If the user needs to configure codegen, for example for Apollo client, the user can still use codegen CLI, but without worrying about Gatsby at all.

d4rekanguok commented on src/gatsby-node.ts:
```diff
@@ -0,0 +1,129 @@
+
+  if (process.env.NODE_ENV === 'development') {
+    const insertTypeWorker = autoFix && setupInsertTypeWorker({ reporter });
+    const watcher = chokidar.watch([
```
in gatsby-plugin-graphql-codegen we're piggybacking on Gatsby's own watcher process by hooking into the redux state directly — not sure if it'll help with the performance

cometkim replied:
Thank you for letting me know! That will be definitely a better option.

I found reducer and state machine about query changing action from here https://github.com/gatsbyjs/gatsby/blob/561d33e2e491d3971cb2a404eec9705a5a493602/packages/gatsby/src/redux/reducers/components.js#L62-L85

All the optimizations I thought are not actually necessary because Gatsby already does! So we can close #26  and #18

d4rekanguok commented on src/gatsby-utils.ts:
```diff
@@ -0,0 +1,113 @@
+  }): Promise<string[]>;
+}
+// Original implementation from https://github.com/gatsbyjs/gatsby/blob/561d33e2e4/packages/gatsby/src/query/query-compiler.js
+export const loadGatsbyFiles: LoadGatsbySourcesFn = async ({
```
I was wondering — why is this function necessary vs. using loadDocuments via @graphql-toolkit?

cometkim replied:
It's becuase the behavior more closer to Gatsby's internal. The plugin goal is codegen for only gatsby so using Gatsby's internal makes sense to ensure documents are actually used by Gatsby.

cometkim replied:
Removed this util.

d4rekanguok commented:
The insert types feature looks to be really cool — if I understand correctly, the lib will import type for query automatically? How does it work?

cometkim committed:
Replace graphql to gatsby/graphql

cometkim committed:
Upgrade Yarn to v2

And add some missing depenencies, Enjoy!

cometkim commented:
> The insert types feature looks to be really cool — if I understand correctly, the lib will import type for query automatically? How does it work?

@d4rekanguok there are few possible options:

Option 1. Analyze imports from source codes, It can also be accomplished with RegExp, or Babel or ESlint. Once imported, it would also able to be reordered by eslint-plugin-import.

Option 2. Generate types as global declaration so It would be not needed.

Option 3. Build a codegen plugin for it. such as [react-apollo plugin](https://graphql-code-generator.com/docs/plugins/typescript-react-apollo). So user writes document seperatly into somewhere like `<root>/src/documents/seo.gatsby.graphql` or `./SEO.gatsby.graphql` (with neer-operation-file), then codegen generate `useSEOQuery()`.

Thought?

cometkim committed:
Make resolver plugins into dependency (not optional)

cometkim committed:
[sec] bump-up serialize-javascript

cometkim committed:
Re-define plugin options and validator

- Reimplementing based on Gatsby internals and getting rid of
graphql-toolkit. So there are options which won't be used.

- Fix to use schema in store directly, why did I need
  gatsby-plugin-extract-schema at first place?

cometkim committed:
Reimplement using gatsby internals

cometkim committed:
Ignore vim settings

cometkim committed:
v0.3.0-alpah.7

cometkim committed:
Bump-up dependencies

cometkim committed:
Remove fixed resolution

cometkim commented:
Hey @d4rekanguok , I just reimplemented plugin.

It less uses graphql-toolkit, no child watcher processes. just subscribe gatsby's redux store like yours.

This way is much optimal and makes code cleaner (removed copies of gatsby). Thank you for your good advice.

cometkim commented:
@d4rekanguok And I think we make a difference in how to provide codegen to users. This plugin is opinionated to hide codegen configuration from users, to only pick and handle Gatsby's documents.

If user want different codegen plugins and options (ex. using react-apollo), then need another codegen process for it. Your plugin may handle this case.

cometkim committed:
WIP: Update README

d4rekanguok commented:
> Option 1. Analyze imports from source codes, It can also be accomplished with RegExp, or Babel or ESlint. Once imported, it would also able to be reordered by eslint-plugin-import.
>
> Option 2. Generate types as global declaration so It would be not needed.
>
> Option 3. Build a codegen plugin for it. such as react-apollo plugin. So user writes document seperatly into somewhere like <root>/src/documents/seo.gatsby.graphql or ./SEO.gatsby.graphql (with neer-operation-file), then codegen generate useSEOQuery().

If I understand correctly, in option 3 user would have to write document in a separate file, which is a no-no for me since it makes components harder to understand (where does this data comes from etc.)

Option 1 seems ok, would user use a cli command to activate the auto import?

Option 2 seems to be the easiest option though I think ~~many~~ a small few people would not like the implicitness.

> It less uses graphql-toolkit, no child watcher processes. just subscribe gatsby's redux store like yours.
>
> This way is much optimal and makes code cleaner (removed copies of gatsby). Thank you for your good advice.

I'm glad it helped 💪!

cometkim committed:
Fix language select condition

cometkim committed:
Change TS Maybe<T> to use undefined instead of null

cometkim committed:
Fix typo

cometkim committed:
Provide context type definition to resolvers plugin

cometkim committed:
Add types for Flow

cometkim committed:
Fix to handle tsx,jsx

cometkim committed:
Fix panic on document name changed

cometkim committed:
Cleanup

cometkim committed:
v0.3.0-alpah.10

cometkim committed:
Revert "Fix panic on document name changed"

This reverts commit 3e43ebc9b619e17d569224ac3262e7efd8708566.

cometkim committed:
Add feature for emitting plugin documents, needed for tooling

cometkim committed:
v0.3.0-alpah.11

cometkim committed:
Fix typo in options

cometkim committed:
Update README

cometkim commented:
I added a new feature `emitPluginDocuments` to help IDE integration (like go to def for plugin fragments)

cometkim commented:
I think this is now ready to ship. Let's publish v1! :rocket:

cometkim committed:
Fix resolver context option
